content/post/2019-05-28.Rmd

```diff
@@ -0,0 +1,322 @@
+
+Finally, `Ops.errors` addresses all the arithmetic operators in a unified way: it computes two derivatives depending on `.Generic` and then propagates the uncertainty using an auxiliary function that implements the Taylor Series Method.
+
+Let's clear the workspace and load the `errors` package to check the complete functionality.
+
+```{r}
+rm(list=ls())
+library(errors)
+
+x <- 1:5 + rnorm(5, sd = 0.01)
+y <- 1:5 + rnorm(5, sd = 0.02)
+errors(x) <- 0.01
+errors(y) <- 0.02
+x; y
+(z <- x / y)
+correl(x, x) # one, cannot be changed
+correl(x, y) # NULL, not defined yet
+correl(x, y) <- runif(length(x), -1, 1)
+correl(x, y)
+covar(x, y)
+(z_correl <- x / y)
+(df <- data.frame(x, 3*x, x^2, sin(x), cumsum(x)))
+format(df, notation="plus-minus")
+```
+
+And much more. This is the complete list of methods implemented in the package:
+
+```{r}
+methods(class="errors")
+```
+
+## Implementing new functionality: new generics
+
+So far, we have explored S3 programming mostly by extending existing generics with methods for a new class of our own. But what if we want to add new functionality? You have probably already guessed that you need to implement your own generics. Let's take a step back and use a simpler classic example to demonstrate this.
+
+```{r}
+# constructors
+circle    <- function(r)    structure(list(r=r),      class="circle")
+rectangle <- function(a, b) structure(list(a=a, b=b), class="rectangle")
+
+# generics
+perimeter <- function(shape) UseMethod("perimeter")
+area      <- function(shape) UseMethod("area")
+
+# methods
+print.circle        <- function(x, ...) with(x, cat("r =", r, "\n"))
+perimeter.circle    <- function(shape)  with(shape, 2 * pi * r)
+area.circle         <- function(shape)  with(shape, pi * r^2)
+
+print.rectangle     <- function(x, ...) with(x, cat("a =", a, ", b =", b, "\n"))
+perimeter.rectangle <- function(shape)  with(shape, 2 * (a + b))
+area.rectangle      <- function(shape)  with(shape, a * b)
+
+# usage example
+(x <- circle(5))
+(y <- rectangle(10, 5))
+perimeter(x)
+perimeter(y)
+area(x)
+area(y)
+```
+
+As you can see, implementing a new generic is as easy as defining a function that calls `UseMethod` with the name of the generic. More advanced uses may, for example, manipulate the input variables before calling `UseMethod`, but this basic template would fit most use cases.
+
+## Combining functionality: inheritance
+
+Inheritance is also possible with S3 programming. Instead of assigning a single class, there may be multiple classes defined as a character vector. When this happens, R dispatches the first class, and subsequent calls to `NextMethod` looks for other methods in the class vector. This means that, if we want our object to resemble parent-child relationships, parent classes must go first in the class vector.
```

Comment:
Tras el 2do punto y seguido pones:

When this happens... calls to...

Y creo que sería (salvo que esté entiendo mal la frase) 

calls NextMethod to look... fíjate en el orden del TO y que la 's' del look desaparece.

Reply:
Sobraba la "s" del "look": *subsequent calls to NextMethod look for something*.